upload/views/helpers.py

```diff
@@ -45,3 +50,45 @@ def get_repository_from_string(
         service=service, repo_identifier=repo_identifier
     )
     return repository
+
+
+def get_repository_and_owner_from_slug_and_commit(
+    slug: str, commitid: str
+) -> tuple[Repository | None, Owner | None]:
+    if "::::" not in slug:
+        return None, None
+
+    owner_username, repository_name = slug.rsplit("::::", 1)
+
+    # formatting for GL subgroups
+    if ":::" in owner_username:
+        owner_username = owner_username.replace(":::", ":")
+
```

Comment:
IMO its cleaner to just to do 1 queryset instead of possibly two, I don't think there would be any significant performance hit with the way these 3 tables are joined. The function can be something like this:
```
queryset = Repository.objects.filter(
    name=repository_name,
    author__username=username,
    commit__commitid=commitid
).select_related('author')

if not queryset.exists() or queryset.count() > 1:
    return None, None
    
repository = queryset.first()
return repository, repository.author

Reply:
I agree, but it would only work if the commit already exists in our db.

Could we argue that this is a stronger match, therefore try this first, and if there's no match try again without commitid?

Reply:
Ah you're right on first call there is no Commit in the DB.. that means in the edge case that we have duplicate Owner/Repo pairs then the true tokenless auth would fail.

The other idea to consider is to require BA and TA upload endpoints to send in a `git_service` value in the request body to identify the git provider if using TT, this way the triplet of owner.username, repo.name, and owner.service will guarantee a one_or_none. And with that you don't need the Commit table nor the commitid from the request. Also no possibility of not supporting TT for duplicate Owner/Repo pairs. This is how the other (non-true) tokenless auth works for BA.